Add option to add event type and event topic to admin panel

pradeepgangwar opened the pull request:
#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes <!-- use `ember test` to run all the tests -->
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

#### Short description of what this resolves:
Currently there is not option to add event types and topic for admins. This causes problem while creating events as we are not able to select types and topics of events. This PR fixes this.

Fixes #1153

pradeepgangwar commented:
Option to add subtopic is left. I am working further on that part.

niranjan94 commented:
@pradeepgangwar is there backend support for this ? Are the APIs ready ?

niranjan94 commented on app/router.js:
```diff
@@ -151,6 +151,11 @@ router.map(function() {
       });
       this.route('translations');
     });
+    this.route('event-settings', function() {
```
Move these to content

```
/admin/content/events/topics
/admin/content/events/sub-topics
/admin/content/events/types
```

niranjan94 replied:
Or even better, you could put all three into just a single page too `/admin/content/events` since the functionality needed isn't complex.

pradeepgangwar replied:
Ok, I will try doing that. This means that form for creating all three things would be in same page? Currently I display them in same page but for new creation everything has different form.

pradeepgangwar replied:
Thanks @niranjan94 I will fix this. :smile:

niranjan94 commented on app/routes/admin/event-settings/index.js:
```diff
@@ -0,0 +1,14 @@
+  },
+  async model() {
+    return {
+      'eventTopics': await this.get('store').query('event-topic', {}),
```
Why quotes for the key ?

pradeepgangwar commented:
@niranjan94 Yes, APIs are ready, in fact I am able to create types and topics. But since event subtopic depends on event topic selected, so that needs a bit more implementation. I am working on that.

pradeepgangwar commented:
@niranjan94 I updated the code and now everything is in a single page with updates handled by modals. But I am having difficulty with subtopics. As subtopics depends on topic so when I select a topic from drop-down, A request is made to server and data is fetched but how do I update the dom to render the fetched data?
![screenshot from 2018-05-24 01-14-58](https://user-images.githubusercontent.com/21259802/40447394-08f464c4-5ef0-11e8-9a24-67f0bcfb8c2e.png)
![screenshot from 2018-05-24 01-15-16](https://user-images.githubusercontent.com/21259802/40447395-0950852e-5ef0-11e8-978d-c05e1be65247.png)

pradeepgangwar commented:
@niranjan94 Everything is working fine. Please review the code.
![screenshot from 2018-05-24 01-14-58](https://user-images.githubusercontent.com/21259802/40510558-b9aa4ee4-5fba-11e8-8b24-f80e48cd1bee.png)
![screenshot from 2018-05-24 01-15-16](https://user-images.githubusercontent.com/21259802/40510560-b9f5c1e4-5fba-11e8-9c3c-ee25a8bafd94.png)

niranjan94 commented on app/templates/admin/content/events.hbs:
```diff
@@ -0,0 +1,124 @@
+  <div class="row">
+    <div class="five wide column">
+      <div class="ui center aligned fluid segment">
+        <h3 class="ui left aligned header">
```
Make these headers one level smaller. `h4`

niranjan94 commented on app/controllers/admin/content/events.js:
```diff
@@ -0,0 +1,133 @@
+        });
+    },
+    async fetchSubTopics(getEventTopic) {
+      return {
```
Why are you returning an object ? You can directly return subTopics right ?

pradeepgangwar replied:
I forgot to remove this code. This is not used right now. Sorry. :sweat_smile:

niranjan94 commented on app/controllers/admin/content/events.js:
```diff
@@ -0,0 +1,133 @@
+
+export default Controller.extend({
+
+  subTopics: computed('model.topic', function() {
```
Remove this computed property. Since the subTopics is loaded asynchronously, use an action that gets triggered when the topic dropdown is changed to make a call to the subtopics and set the subtopic once the promise is resolved.

pradeepgangwar replied:
Ok, I will try doing this. :)

pradeepgangwar commented:
@niranjan94 Please check the new UI. And about removing that computed property, I got confused about how to send the event id (event subtopic need event ID for query) to action by simultaneously picking up selected id from dropdown and calling an action. 
I got this implementation from the event wizard where we have implemented this functionality like this. Let me know how this can be done.

![screenshot from 2018-05-26 00-32-48](https://user-images.githubusercontent.com/21259802/40562026-03ec48a0-607d-11e8-8ae0-5be1d0ad5752.png)

niranjan94 commented on app/templates/admin/content/events.hbs:
```diff
@@ -0,0 +1,157 @@
+        <div class="ui divider"></div>
+        <div class="field">
+          <label>{{t 'Event Topic'}}</label>
+          {{#ui-dropdown class='search selection' onChange=(action (mut model.topic)) as |execute mapper|}}
```
here instead of `(action (mut model.topic))` make it something like `(action 'loadSubTopic')`.

niranjan94 replied:
here instead of `(action (mut model.topic))` make it something like `(action 'loadSubTopic')`.

Then define an action like, 

```js
async loadSubTopic(topic) {
  this.set('subTopics', await topic.get('subTopics');
}
``` 

Make sure to change the implementation in the new subtopic modal too. Also, ensure you have a loading indicator that comes up when the subtopics are loading.

niranjan94 commented on app/controllers/admin/content/events.js:
```diff
@@ -0,0 +1,128 @@
+    return this.get('model.topic.subTopics');
+  }),
+
+  actions: {
```
Too much repetition of very similar stuff. 

`openNewEventTypeModal`, `openNewEventTopicModal`, `openNewEventSubTopicModal` can become one action
`deleteEventType`, `deleteEventTopic`, `deleteEventSubTopic` can become one action
`addEventType`, `addEventTopic`, `addEventSubTopic` can become one action.

pradeepgangwar commented:
@niranjan94 I removed the computed property and added tests. However in the actions I was able to reduce delete type action but could not reduce add and openModal type actions. Those actions involve different calls and parameters each time. I tried different ways and one way was to add a key value pair in models of each event-type, event-topic and event-sub-topic that will identify which parameter was sent and then choose action. Although that didn't reduce code that much. Is that the correct way?

niranjan94 commented on app/controllers/admin/content/events.js:
```diff
@@ -0,0 +1,95 @@
+        this.set('isLoading', false);
+      }
+    },
+    openNewEventTypeModal(eventType) {
```
The three can be,

```js
function openModalFor(modelInstanceOrName) {
  const modelName = typeof modelInstanceOrName === 'string' ? modelInstanceOrName : modelInstanceOrName.modelName;
  const modelInstance = typeof modelInstanceOrName === 'string' ? this.store.createRecord(modelInstanceOrName) : modelInstanceOrName;
  const camelCasedValue = camelCase(modelName); // from lodash
  this.set(camelCasedValue, modelInstance); // You can use this property itself to decide whether the model should be visible or not since `null` ==> falsy
}
```

niranjan94 commented on app/controllers/admin/content/events.js:
```diff
@@ -0,0 +1,95 @@
+          this.set('isLoading', false);
+        });
+    },
+    addEventType() {
```
All these add actions can just be 

```js
addEventProperty(modelInstance) {
  const camelCasedValue = camelCase(modelInstance.modelName);
  this.set('isLoading', true);
  modelInstance.save()
    .then(() => {
      this.notify.success(this.get('l10n').t(`${startCase(camelCasedValue)} has been added successfully.`));
    })
    .catch(() => {
      this.notify.error(this.get('l10n').t(`An unexpected error has occurred. ${startCase(camelCasedValue)} not saved.`));

    })
    .finally(() => {
      this.set(camelCasedValue, null);
      this.set('isLoading', false);
    })

}
```

And you can pass in the `modelInstance` in the template.

niranjan94 commented on app/components/modals/admin/content/new-event-sub-topic-modal.js:
```diff
@@ -0,0 +1,49 @@
+  isSmall            : true,
+  autoScrollToErrors : false,
+
+  eventSubTopic: computed('model.topic', function() {
```
This is still pending. Remove usage of a computed property for this.

pradeepgangwar commented:
@niranjan94 Everything is working and code has been reduced. :smile: I was stuck at previous step for long but your solution was amazing and great. :smile: A good learning for me.

pradeepgangwar commented:
@niranjan94 Everything is working and code has been reduced. smile I was stuck at previous step for long but your solution was amazing. smile A good learning for me.
Please review the code.

niranjan94 commented on app/controllers/admin/content/events.js:
```diff
@@ -0,0 +1,73 @@
+      this.set(camelCasedValue, modelInstance);
+      switch (camelCasedValue) {
+        case 'eventType':
+          this.set('isNewEventTypeModalOpen', true);
```
Like I said before, these properties (`isNewEventTypeModalOpen` and such) are not needed. You can just directly pass in `eventType` to `isOpen`. That would be enough to control the model.

pradeepgangwar commented:
@niranjan94 did the changes. Please review.
I needed to add a openModal attribute to model instance. Without that it as giving me an error which says. `Assertion Failed: Cannot call set with 'name' on an undefined object` But now it works fine.

pradeepgangwar commented:
@niranjan94 did the changes. Please review.
I needed to add a openModal attribute to model instance. Without that it was giving me an error which says. Assertion Failed: Cannot call set with 'name' on an undefined object But now it works fine.

pradeepgangwar commented:
@niranjan94 Sorry for pushing my changes again, there was some problem with the edit button. I had to replace modelName with `constructor.modelName` to make it work. I didn't check that earlier. Everything seems to work now.
Please review again. :sweat_smile:

ritikamotwani commented on app/components/modals/admin/content/new-event-sub-topic-modal.js:
```diff
@@ -0,0 +1,33 @@
+      delay  : false,
+      on     : 'blur',
+      fields : {
+        email: {
```
@pradeepgangwar why is it named email when the field is a name.

pradeepgangwar commented:
@niranjan94 @ritikamotwani Sorry for trouble again. Please review.
